app/upgrades/v11/upgrades.go

```diff
@@ -74,3 +94,78 @@ func MigrateEscrowAccounts(ctx sdk.Context, ak authkeeper.AccountKeeper) {
 		ak.SetModuleAccount(ctx, acc)
 	}
 }
+
+// HandleRewardDistribution handles the logic for the reward distribution, it only commits to the db if successful
+func HandleRewardDistribution(ctx sdk.Context, bk bankkeeper.Keeper, sk stakingkeeper.Keeper, dk distributionkeeper.Keeper, logger log.Logger) {
+	// use a cache context as a rollback mechanism in case
+	// the distrbution fails
+	cacheCtx, writeFn := ctx.CacheContext()
+	err := DistributeRewards(cacheCtx, bk, sk, dk)
+	if err != nil {
+		// log error instead of aborting the upgrade
+		logger.Error("failed to distribute rewards", "error", err.Error())
+	} else {
+		writeFn()
+	}
+}
+
+// DistributeRewards distributes the token allocations from the Olympus Mons incentivized testnet
+func DistributeRewards(ctx sdk.Context, bk bankkeeper.Keeper, sk stakingkeeper.Keeper, dk distributionkeeper.Keeper) error {
+	funder := sdk.MustAccAddressFromBech32(FundingAccount)
+	numValidators := sdk.NewInt(int64(len(Validators)))
+
+	for _, allocation := range Allocations {
+
+		// send rewards to receivers
+		receiver := sdk.MustAccAddressFromBech32(allocation[0])
+		receivingAmount, ok := sdk.NewIntFromString(allocation[1])
+		if !ok {
+			return errorsmod.Wrapf(errortypes.ErrInvalidType,
+				"cannot retrieve allocation from string for address %s",
+				allocation[0])
+		}
+		reward := sdk.Coins{
+			sdk.NewCoin(types.BaseDenom, receivingAmount),
+		}
+		err := bk.SendCoins(ctx, funder, receiver, reward)
+		if err != nil {
+			return err
+		}
+
+		// delegate receiver's rewards to validators selected validators equally
+		delegationAmt := reward.QuoInt(numValidators)[0].Amount
+		remainderAmount := (reward[0].Amount).Mod(numValidators)
+		for i, validatorBech32 := range Validators {
+			validatorAddress, err := sdk.ValAddressFromBech32(validatorBech32)
+			if err != nil {
+				return err
+			}
+			validator, found := sk.GetValidator(ctx, validatorAddress)
+			if !found {
+				return errorsmod.Wrapf(slashingtypes.ErrBadValidatorAddr,
+					"validator address %s cannot be found",
+					validatorAddress)
+			}
+			// we delegate the remainder to the first validator, for the sake of testing consistency
+			// this remainder is in the order of 10^-18 evmos, and at most 10^-15 evmos after all rewards are allocated
+			if remainderAmount.IsPositive() && i == 0 {
```

Comment:
Instead of calling Delegate twice you can write

```suggestion
			if remainderAmount.IsPositive() && i == 0 {
			  delegatioAmt = delegationAmt.Add(remainderAmount)
			}
			_, err = sk.Delegate(ctx, receiver, delegationAmt, 1, validator, true)
```

Reply:
If we reassign delegationAmt, it makes it so the delegation amount changes on future iterations for the loop, and we only want it to change for the first validator.

Reply:
This will introduce a bug. It will update the `delegationAmt` and in case of `remainderAmount > 0` will delegate more funds that available for the following validators (`i > 0`)